File: backend/internal/api/handler/stock_symbols.go
package handler

import (
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"

	"github.com/trading/backend/internal/store"
	"github.com/trading/backend/pkg/timeutil"
)

type StockSymbolHandler struct {
	repo store.StockSymbolRepo
	log  *zap.Logger
}

func NewStockSymbolHandler(repo store.StockSymbolRepo, log *zap.Logger) *StockSymbolHandler {
	return &StockSymbolHandler{repo: repo, log: log}
}

// Candidates 產生研究用的候選標的清單（GET /api/v1/stock-symbols/candidates）。
//
// 用途是 todo.md T-040 的擴標的池：Step 1 要「全部 ETF ＋ 各產業分層抽樣約 300 檔股票」，
// 手工湊 650 個代號不切實際。回傳的 symbols 可直接餵給 POST /api/v1/market/backfill。
//
// Query：
//
//	security_type=股票,ETF   逗號分隔。**留空預設 `股票,ETF`**，見下方說明
//	industry=半導體業,航運業  逗號分隔，空 = 不限
//	listed_years=5           只留上市滿 N 年的（listed_date 為 NULL 者一律排除）。0 或留空 = 不限
//	per_industry=9           每個產業最多幾檔（在該產業代號區間內等距取樣；半導體業有 201 檔，不設限會被它主導）
//	limit=650                總筆數上限，預設 3000、上限 5000
//	include_delisted=true    預設 false
//
// **security_type 為什麼有預設值**：`stock_symbols` 存的是完整的 TWSE ISIN 主檔，
// 實測 43,061 筆上市資料裡有 40,658 筆是認購（售）權證——佔 94%，而且代號排序在股票之前。
// 不給預設值的話，一個沒帶參數的請求會回傳「ETF ＋ 權證」而一檔股票都沒有，
// 且這份清單被設計成可直接餵給 `POST /market/backfill`（無筆數上限、5 req/min），
// 等於把數小時的 FinMind 配額花在沒有 K 線的商品上。要權證請明確指定。
func (h *StockSymbolHandler) Candidates(c *gin.Context) {
	securityTypes := splitCSVParam(c.Query("security_type"))
	if len(securityTypes) == 0 {
		securityTypes = defaultCandidateSecurityTypes
	}
	opts := store.StockSymbolCandidateOptions{
		SecurityTypes: securityTypes,
		Industries:    splitCSVParam(c.Query("industry")),
	}

	if raw := strings.TrimSpace(c.Query("listed_years")); raw != "" {
		years, err := strconv.Atoi(raw)
		if err != nil || years < 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "listed_years must be a non-negative integer"})
			return
		}
		// years == 0 視為「不限制」而不是「上市日 <= 現在」。後者會連帶啟用
		// `listed_date IS NOT NULL` 而靜靜濾掉解析失敗的列（parseListedDate 對格式錯誤
		// 只記 warning 不中斷），而前端數字輸入框的預設值很容易就是 0。
		if years > 0 {
			opts.ListedBefore = time.Now().In(timeutil.TaipeiTZ).AddDate(-years, 0, 0)
		}
	}
	if raw := strings.TrimSpace(c.Query("per_industry")); raw != "" {
		n, err := strconv.Atoi(raw)
		if err != nil || n < 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "per_industry must be a non-negative integer"})
			return
		}
		// 0 視為「不限制」，理由同上面的 listed_years：數字輸入框被清空或歸零時會送 0，
		// 那應該回傳不設限的清單，而不是整個請求 400。
		opts.PerIndustryLimit = n
	}
	if raw := strings.TrimSpace(c.Query("limit")); raw != "" {
		n, err := strconv.Atoi(raw)
		if err != nil || n <= 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "limit must be a positive integer"})
			return
		}
		opts.Limit = n
	}
	if raw := strings.TrimSpace(c.Query("include_delisted")); raw != "" {
		parsed, err := strconv.ParseBool(raw)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "include_delisted must be true or false"})
			return
		}
		opts.IncludeDelisted = parsed
	}

	result, err := h.repo.ListCandidates(c.Request.Context(), opts)
	if err != nil {
		serverError(c, h.log, err, "stock symbols: candidates")
		return
	}

	// symbols 是給 market/backfill 直接用的扁平清單；rows 保留完整欄位供人工核對產業分佈。
	symbols := make([]string, 0, len(result.Symbols))
	byIndustry := map[string]int{}
	for _, row := range result.Symbols {
		symbols = append(symbols, row.Symbol)
		byIndustry[row.Industry]++
	}
	c.JSON(http.StatusOK, gin.H{
		"count":       len(result.Symbols),
		"symbols":     symbols,
		"by_industry": byIndustry,
		"rows":        result.Symbols,
		// truncated=true 代表還有更多符合條件的標的被 limit 砍掉，而截斷是依代號順序，
		// 會整批砍掉高代號的產業——呼叫端要知道這份清單不完整。
		"truncated": result.Truncated,
	})
}

// 研究母體預設只看股票與 ETF；權證等其他 ISIN 類別要明確指定才會出現。
var defaultCandidateSecurityTypes = []string{"股票", "ETF"}

func splitCSVParam(raw string) []string {
	parts := strings.Split(raw, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if p = strings.TrimSpace(p); p != "" {
			out = append(out, p)
		}
	}
	return out
}

func (h *StockSymbolHandler) Search(c *gin.Context) {
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))
	onlyListed := true
	if raw := strings.TrimSpace(c.Query("listed")); raw != "" {
		parsed, err := strconv.ParseBool(raw)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "listed must be true or false"})
			return
		}
		onlyListed = parsed
	}

	rows, err := h.repo.Search(c.Request.Context(), store.StockSymbolSearchOptions{
		Query:        c.Query("q"),
		OnlyListed:   onlyListed,
		SecurityType: c.Query("security_type"),
		Limit:        limit,
	})
	if err != nil {
		serverError(c, h.log, err, "stock symbols: search")
		return
	}
	c.JSON(http.StatusOK, gin.H{"symbols": rows})
}

File: frontend/src/lib/api/stockSymbols.ts
import { apiFetch } from './client'
import type { StockSymbol } from './watchlist'

// 研究用的候選標的清單（後端 GET /stock-symbols/candidates，見 docs/todo.md T-040）。
//
// **為什麼另開一個模組**：`/stock-symbols/search` 借放在 watchlist.ts，因為那支端點的用途
// 就是 watchlist 新增股票時的 autocomplete。候選清單與 watchlist 無關——它是研究母體的
// 產生器，回傳的 symbols 直接餵給 market/backfill。放同一個檔案會讓兩種用途互相牽制。
// `StockSymbol` 型別本身是共用的，從 watchlist.ts re-export 而不是複製一份。
export type { StockSymbol }

export interface SymbolCandidateOptions {
  /** 留空時**後端**會套用預設 `股票,ETF`；前端不重複實作這個預設值。 */
  securityTypes?: string[]
  industries?: string[]
  /** 上市滿幾年。0 或 undefined = 不限。 */
  listedYears?: number
  /** 每個產業最多幾檔。0 或 undefined = 不限。 */
  perIndustry?: number
  limit?: number
  includeDelisted?: boolean
}

export interface SymbolCandidatesResult {
  count: number
  /** 扁平代號清單，可直接餵給 triggerBackfill。 */
  symbols: string[]
  /** 產業 → 取樣檔數。用來人工核對抽樣有沒有被單一產業壓垮。 */
  by_industry: Record<string, number>
  rows: StockSymbol[]
  /**
   * true 代表還有更多符合條件的標的被 limit 砍掉。**截斷依代號順序**，會整批砍掉
   * 高代號的產業，正是 per_industry 要消除的偏斜——拿到 true 時應該調高 limit
   * 或收緊條件，而不是直接使用這份清單。
   */
  truncated: boolean
}

/** 篩選選項與**母體**筆數（不是取樣後的數量）。 */
export interface SymbolFacet {
  value: string
  count: number
}

export interface SymbolFacets {
  security_types: SymbolFacet[]
  industries: SymbolFacet[]
}

/**
 * 取得可選的證券類型與產業。
 *
 * `securityTypes` 只縮放回傳的 industries，**不影響 security_types 清單**——
 * 選單本身要一直完整，否則使用者選了某個類型之後就換不回來。
 */
export async function fetchSymbolFacets(securityTypes?: string[]): Promise<SymbolFacets> {
  const params = new URLSearchParams()
  if (securityTypes?.length) params.set('security_type', securityTypes.join(','))
  const qs = params.toString()
  const res = await apiFetch<SymbolFacets>(`/stock-symbols/facets${qs ? `?${qs}` : ''}`)
  return {
    security_types: res.security_types ?? [],
    industries: res.industries ?? [],
  }
}

export async function fetchSymbolCandidates(
  opts: SymbolCandidateOptions = {},
): Promise<SymbolCandidatesResult> {
  const params = new URLSearchParams()
  if (opts.securityTypes?.length) params.set('security_type', opts.securityTypes.join(','))
  if (opts.industries?.length) params.set('industry', opts.industries.join(','))
  // 0 不送出：後端把 0 當成「不限制」，但不送更明確，也少一個會漂移的約定。
  if (opts.listedYears) params.set('listed_years', String(opts.listedYears))
  if (opts.perIndustry) params.set('per_industry', String(opts.perIndustry))
  if (opts.limit) params.set('limit', String(opts.limit))
  if (opts.includeDelisted) params.set('include_delisted', 'true')

  const qs = params.toString()
  const res = await apiFetch<SymbolCandidatesResult>(
    `/stock-symbols/candidates${qs ? `?${qs}` : ''}`,
  )
  return {
    count: res.count ?? 0,
    symbols: res.symbols ?? [],
    by_industry: res.by_industry ?? {},
    rows: res.rows ?? [],
    truncated: res.truncated ?? false,
  }
}
